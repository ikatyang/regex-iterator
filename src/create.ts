/**
 * create a generator for string matching
 */
export function* create(regex: RegExp, str: string, max_result = Infinity) {
  const cloned_regex = new RegExp(regex.source, regex.flags);

  let counter = 0;
  let result = cloned_regex.exec(str);

  while (counter < max_result && result !== null) {
    counter++;

    yield result;

    if (!regex.global) {
      break;
    }

    result = cloned_regex.exec(str);
  }
}
