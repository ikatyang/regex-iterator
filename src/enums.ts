export const enum TakeAction {
  Break = -1,
  TakeAndBreak = 0,
  TakeAndContinue = 1,
}
