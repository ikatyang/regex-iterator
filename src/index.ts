export * from './create';
export * from './enums';
export * from './iterate';
export * from './take';
