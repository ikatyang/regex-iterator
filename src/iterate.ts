import {TakeAction} from './enums';

/**
 * iterate the generator
 */
export const iterate = <T>(
    iterator: IterableIterator<T>,
    fn: (value: T, index: number) => void | boolean | TakeAction,
    ) => {

  let counter = 0;
  let is_continue = true;
  let result = iterator.next();

  while (is_continue && !result.done) {

    const action = fn(result.value, counter++);

    switch (action) {
      case false:
      case TakeAction.Break:
      case TakeAction.TakeAndBreak:
        is_continue = false;
        break;

      default:
        is_continue = true;
        break;
    }

    if (is_continue) {
      result = iterator.next();
    }
  }
};
