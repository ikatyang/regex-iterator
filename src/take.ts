import {TakeAction} from './enums';
import {iterate} from './iterate';

/**
 * take items in the generator
 */
export const take = <T, U = T>(
    iterator: IterableIterator<T>,
    pred: (value: T) => void | boolean | TakeAction,
    fn: (value: T, index: number, values: U[]) => U = (x: T) => x as any as U,
    ): U[] => {
  const values: U[] = [];

  iterate(iterator, (value, index) => {
    const action = pred(value);

    switch (action) {

      case true:
      case TakeAction.TakeAndBreak:
      case TakeAction.TakeAndContinue:
        values.push(fn(value, index, values));
        break;

      default:
        // do nothing
        break;
    }

    return action;
  });

  return values;
};
