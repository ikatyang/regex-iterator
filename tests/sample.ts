import * as regex_iterator from '../src/index';

// tslint:disable:comment-format

const spread_sample = () => {
  const iterator = regex_iterator.create(/([0-9])([a-z])/g, '1a 2b 3c');
  return [...iterator]; //=> [['1a', '1', 'a'], ['2b', '2', 'b'], ['3c', '3', 'c']]
};

const take_sample = () => {
  const iterator = regex_iterator.create(/([0-9])([a-z])/g, '1a 2b 3c');
  return regex_iterator.take(
    iterator,
    match => match[2] !== 'c',
    match => `${match[1]}-${match[2]}`,
  ); //=> ['1-a', '2-b']
};

it('should spread correctly', () => {
  expect(spread_sample()).toMatchSnapshot();
});

it('should take correctly', () => {
  expect(take_sample()).toMatchSnapshot();
});
