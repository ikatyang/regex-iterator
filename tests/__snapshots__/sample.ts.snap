// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should spread correctly 1`] = `
Array [
  Array [
    "1a",
    "1",
    "a",
  ],
  Array [
    "2b",
    "2",
    "b",
  ],
  Array [
    "3c",
    "3",
    "c",
  ],
]
`;

exports[`should take correctly 1`] = `
Array [
  "1-a",
  "2-b",
]
`;
